update_ip/services/cloudflare.py

```diff
@@ -0,0 +1,178 @@
+        self.dns_record = domain
+
+    def parse_result(self, data):
+        """Parse the JSON result and return a dict/list with data
+        :param data: JSON data to parse
+        """
+        return json.loads(data)
+
+    def update_ip(self, ip, second_try=False):
+        """Get the DNS Record info for selected sub-domain
+        :param ip: IP address to update
+        :param second_try: If Record_ID is empty try to get it, if everything fails in a second attempt, we throw an
+                            exception.
+        """
+        if self.record_id is None:
+            if second_try:
+                raise DNSServiceError(
+                    'Record ID is empty, we tried to find it, but it was not possible'
+                )
+            self.get_dns_record()
+            self.update_ip(ip, True)
+            return
+
+        self.create_request(self.generate_update_url(),
+                            json.dumps({'name': self.dns_record, 'type': self.record_type, 'content': ip})
+                            )
+
+        self.request.get_method = lambda: 'PUT'
+        result = self.opener.open(self.request)
+        pprint(result)
+        data = self.parse_result(result.read())
+        pprint(data)
+
+    def get_zones(self):
+        """Get the zone info for selected domain
+        """
+        self.create_request(self.generate_zones_list_url())
+        result = urllib2.urlopen(self.request)
+        data = self.parse_result(result.read())
+        if len(data[u'result']) > 0:
+            self.zone_id = data[u'result'][0][u'id']
+        else:
+            raise DNSServiceError('No info found for domain: ' + self.domain)
+
+    def get_dns_record(self, second_try=False):
+        """Get the DNS Record info for selected sub-domain
+        :param second_try: If Zone_ID is empty try to get it, if everything fails in a second attempt, we throw an
+                            exception.
+        """
+        if self.zone_id is None:
+            if second_try:
+                raise DNSServiceError(
+                    'Zone ID is empty, we tried to find it, but it was not possible'
+                )
+            self.get_zones()
+            self.get_dns_record(True)
+            return
+
+        self.create_request(self.generate_dns_list_url())
+        result = urllib2.urlopen(self.request)
+        data = self.parse_result(result.read())
+
+        if len(data[u'result']) > 0:
+            self.record_id = data[u'result'][0][u'id']
+        else:
+            raise DNSServiceError(
+                'No info found for DNS Record: "' + self.dns_record + '" with DNS type: "' + self.record_type + '"'
+            )
+
+    def create_request(self, url, data=None):
+        """Creates the HTTP request and add headers
+        :param url: API URL with the petition.
+        :param data: Request Data
+        """
+        if data is None:
+            self.request = urllib2.Request(url)
```

Comment:
What I meant is that, unless there is a good reason not to, we should get rid of `self.request` *entirely* - in all methods. We shouldn't keep internal state in the class instance if we don't need to, since it can break unanticipated use cases